### PR TITLE
Initial implementation for ruby autocorrection

### PR DESCRIPTION
**Pull request by EiNSTeiN-**
This implements autocorrection for violations found by the Ruby linter. The idea is that users of this gem can call  `autocorrect(file)` to auto-correct all modified lines in a PR. We can then present the offending line and the corrected line together for example, or write a bit of code to push corrected lines to github (this might be outside the scope of this gem though).

This duplicates some code form rubocop's `Team` class because unlike rubocop, policial only applies corrections that are found on lines that have changed, and this isn't possible at all with rubocop. The code to make this work comes from here: https://github.com/bbatsov/rubocop/blob/de1b1e037efe4c9d30b0319de4ab2e55e583cc25/lib/rubocop/cop/team.rb#L85-L143 and it was adapted to work in the context of policial.

@volmer

### CODE REVIEW

File: lib/policial/linters/ruby.rb [hunk 1]
```diff
@@ -32,22 +48,72 @@ def default_config_file
...
                         RuboCop::Cop::Cop.non_rails
                       end
 
-        RuboCop::Cop::Team.new(cop_classes, config, extra_details: true)
+        RuboCop::Cop::Team.new(
+          cop_classes,
+          config,
+          extra_details: true,
+          auto_correct: auto_correct,
+          stdin: ''
```
Comment (EiNSTeiN-) [inline lib/policial/linters/ruby.rb:110]: Unless `stdin` is passed as an option, rubocop will try to save the modified file on disk when `auto_correct: true`...

File: lib/policial/linters/ruby.rb [hunk 2]
```diff
@@ -32,22 +48,72 @@ def default_config_file
...
+
+      def collate_corrections(file, corrector, team)
+        skips = Set.new
+        team.cops.each do |cop|
+          next if cop.corrections.empty? || skips.include?(cop.class)
+
+          cop.offenses.select(&:corrected?).each_with_index do |offense, index|
+            next unless build_violation(file, offense).on_changed_line?
+
+            corrector.corrections << cop.corrections[index]
```
Comment (EiNSTeiN-) [inline lib/policial/linters/ruby.rb:92]: When the team is run with `auto_correct: true`, corrections are inserted in `cop.corrections` whenever an offense is corrected, so we need to iterate over corrected cops only or indexes will not match, hence the use of `cop.offenses.select(&:corrected?)`

**Comment by volmer** [comment]
I like the idea! And I also agree that pushing the actual change back to the PR is outside of the scope of this gem, but doable in the hosting app.

So is the goal to `autocorrect` to return the entire content of the file with the corrections performed or just the correction for specific violations? Would the hosting app have to call `autocorrect` for each file?

**Comment by EiNSTeiN-** [comment]
> So is the goal to autocorrect to return the entire content of the file with the corrections performed or just the correction for specific violations?

The implementation in this PR does the former, it returns the contents of the autocorrected file.

> Would the hosting app have to call autocorrect for each file?

Yes, it does one file at a time, matching how `violations_in_file` works

**Comment by EiNSTeiN-** [comment]
> We can then present the offending line and the corrected line together

After thinking about this more, it would be tricky to match the source lines with the corrected lines because the correction process can insert or delete lines in the corrected file. Doing that would require keeping a copy of the changes that each correction pushes into the [`Parser::Source::Rewriter#active_queue`](https://github.com/whitequark/parser/blob/master/lib/parser/source/rewriter.rb). This might be good future work, for now the content of the autocorrected file is still valuable for the host app, we'll just need to show the entire file instead of specific lines.

**Comment by volmer** [comment]
Had to face a bunch of conflicts so I ended up cherry-picking some commits and reimplementing the code on top of the latest `master`, based on this PR. Changes were applied starting at 3545b25adadeff7a637cd21e1c1f8d75ebe5d1b9
 It works wonderfully, all credits to you @EiNSTeiN- Thank you!

**Pull request closed without merging**